Feature/basicwlfinalization

NikolaiMe opened the pull request:
Added BasicWL support plus tests

NikolaiMe committed:
Prepare BasicWL XML scheme

NikolaiMe committed:
Bugfix: Correct BasicWL XML Profile to fit requirements

NikolaiMe committed:
Feature BasicWL Profile added

NikolaiMe committed:
BasicWL Add mapping

NikolaiMe committed:
Add Tests for basicwl
- Additionally: Small adaption of the converter classes to allow easier debugging: abstract function are now _toValue() and _toXML(). In BasicConverter class those functions are now called in non-abstract functions toValue() and toXML() --> this allows to easily log out some data and is very helpful when some conversion failed (and you don't know exactly where the conversion failed)

NikolaiMe committed:
adapt naming of basicwl_test_objects.ts

schwarmco commented on package.json:
```diff
@@ -58,6 +58,7 @@
         "typescript-eslint": "^8.15.0",
-        "xmllint-wasm": "^4.0.2"
+        "xmllint-wasm": "^4.0.2",
+        "zod-to-ts": "^1.2.0"
```
is `zod-to-ts` already in use afar from the playground-test?

NikolaiMe replied:
correct I use it to create and print-out the full ts-type. It helps a lot when the XSD validation fails and I need to find out what's the bug in our profile. And I do this only via playground tests. But I don't use it somewhere else.
Btw. I basically only have this playground test because 'npx ts-node whateverfile.ts' is not working in our setup. It always tells me that .ts is a unknown file-format, which is quite weird because it's **ts**-node... And in other projects this works perfectly. Do you have an idea why this is not working? I already tried to fix it by adapting the ts-config, but this messed up everything. So I had this workaround to execute the code via a test...

NikolaiMe replied:
Do you want me to take any actions for this? Shall I remove the dependency (i would rather keep it, because I will need it in the future again)

schwarmco replied:
no, thats fine if we're going to use it - i will look into the other issue with ts-node you mentioned though

schwarmco commented on src/profiles/convert.ts:
```diff
@@ -2,64 +2,84 @@
+    ? {
+          [K in keyof T & string]:
+              | K
+              | (NonNullable<T[K]> extends (infer U)[]
```
that's clever!

NikolaiMe replied:
Yeah more or less clever :-D ... this is the result of one hour researching, trial and error and useless Github Co-Pilot discussions. But finally it worked and didn't mess up the optional keys anymore.

schwarmco commented on src/profiles/convert.ts:
```diff
@@ -84,7 +104,7 @@ export abstract class Converter<Profile, ProfileXml> {
             }
 
-            if (item.arrayMap && Array.isArray(value)) {
+            if (Array.isArray(value) && !(item.converter instanceof SpecifiedTaxRegistrationsForSellerTypeConverter)) {
```
that's a very specific edge case check in the converter here - is this guaranteed to only be the case for the sellers tax numbers or could this affect other converters too (i.e. in extended profile)? 

if so, what do you think about having a flag in the converter that specifies this, i.e.:

```ts
if (Array.isArray(value) && item.converter.isArrayable) {
```

NikolaiMe replied:
Yeah this was a quick and dirty solution for the problem, that this single converter takes an array and maps the items of that array to two keys in an object.
I didn't check whether this might be an issue in the future, too tbh....

Wait a second I just found a bug. Because when talking about that "single converter that takes an array" I remembered that TradeAllowanceChargeTypeConverter also takes an Array, and I didn't handle that yet, and the tests do not check this yet, as the test-pdf does not have allowances or charges.

--> Long story short: We should use your implementation because we already have two of those edge-cases already.

Plus: I don't like it, that the arrays in "our-schemes" are unions right now (either object or array). This makes the handling of the data unnecessarily hard (because you always need to check, is it an array or is it an object because you can use your data). I would prefer to always have an array instead... And I think we could also use your solution to make this happen...

I try to implement this.

NikolaiMe replied:
I implemented it a little bit differently. Because my solution solves both issues (specific excludes in convert.ts and unions object | object[] removed from object scheme). Plus I even more embrace the Converter logic, which I think is more consistent than the old implementation.

See details below

NikolaiMe committed:
Refactor | Change Array Handling
--> Before Refactoring Arrays were handled in convert.ts. This had some ugly sideeffects, like:
- Converters which expect an array to be passed to the converters need to be excluded from the handling in convert.ts
- The object scheme needed to handle both, objects (when the 'array' only had one element) and arrays. This made the object scheme unnecessarily complex.

--> After refactoring: There is now a "super-converter" for arrays. Now the elements in the mapping which can be arrays need to wrap their converters in an array Converter like: new ArrayConverter(new IdTypeConverter()).
Now, if there could be more then one item for an key in the object scheme, the value will always be an array (if there's only one item, it's an array with one item instead of a plain object).
Plus the ugly handling in the convert.ts is gone.

NikolaiMe committed:
Cleared up some ts linting errors because of unused imports
